src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -1451,60 +1451,33 @@ internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnota
             return this.WithUnderlyingType(underlyingType);
         }
 
-        internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind variance)
+        internal override TypeSymbol MergeEquivalentTypes(TypeSymbol other, VarianceKind variance)
         {
             Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
-            var otherTuple = other as TupleTypeSymbol;
-            if (otherTuple is null)
+            var mergedUnderlying = (NamedTypeSymbol)this.TupleUnderlyingType.MergeEquivalentTypes(other.TupleUnderlyingType, variance);
+
+            ImmutableArray<string> names1 = this._elementNames;
+            ImmutableArray<string> names2 = ((TupleTypeSymbol)other)._elementNames;
+            ImmutableArray<string> mergedNames;
+            if (names1.IsDefault || names2.IsDefault)
             {
-                return this;
+                mergedNames = default;
             }
-            NamedTypeSymbol underlyingType;
-            if (MergeUnderlyingTypeNullability(_underlyingType, otherTuple._underlyingType, variance, out underlyingType))
+            else
             {
-                return WithUnderlyingType(underlyingType);
-            }
-            return this;
-        }
-
-        private static bool MergeUnderlyingTypeNullability(
-            NamedTypeSymbol typeA,
-            NamedTypeSymbol typeB,
-            VarianceKind variance,
-            out NamedTypeSymbol mergedType)
-        {
-            Debug.Assert(typeA.Equals(typeB, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
+                Debug.Assert(names1.Length == names2.Length);
+                mergedNames = names1.ZipAsArray(names2, (n1, n2) => string.CompareOrdinal(n1, n2) == 0 ? n1 : null);
```

Comment:
I don't know how much you want to care about this, but this will force us to do two full enumerations of the names list, as we do one here, and then one below when we determine if the merged names should be default. We then do a 3rd iteration in the sequence equal, when that really only needs to be done in this branch. Consider rewriting as a simple for loop, and making flags for the conditions being checked in the later iterations.

Reply:
I tried the change as you suggested, and it isn't clear to me that it is worth it (it seems more fragile as it has more invariants across larger regions of code), so I'm keeping the existing implementation.

---
In reply to: [332124789](https://github.com/dotnet/roslyn/pull/39069#discussion_r332124789) [](ancestors = 332124789)